File: src/data/courses/dsa/graphs/traversals-basic-properties/check-bipartite/code/java/CheckBipartite.java
// ============================================================
// Check Bipartite
// ============================================================
import java.util.*;
class Solution {
    public boolean isBipartite(int[][] adj) {
        int n = adj.length;
        int[] color = new int[n]; Arrays.fill(color, -1);
        for (int i = 0; i < n; i++) {
            if (color[i] != -1) continue;
            Queue<Integer> q = new LinkedList<>();
            q.add(i); color[i] = 0;
            while (!q.isEmpty()) {
                int node = q.poll();
                for (int nb : adj[node]) {
                    if (color[nb] == -1) { color[nb] = 1-color[node]; q.add(nb); }
                    else if (color[nb] == color[node]) return false;
                }
            }
        }
        return true;
    }
}

File: src/data/courses/dsa/graphs/traversals-basic-properties/check-bipartite/code/python/check_bipartite.py
# ============================================================
# Check Bipartite
# ============================================================
from collections import deque
class Solution:
    def isBipartite(self, adj):
        n = len(adj)
        color = [-1] * n
        for i in range(n):
            if color[i] != -1: continue
            queue = deque([i]); color[i] = 0
            while queue:
                node = queue.popleft()
                for nb in adj[node]:
                    if color[nb] == -1:
                        color[nb] = 1 - color[node]
                        queue.append(nb)
                    elif color[nb] == color[node]:
                        return False
        return True
